sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala

```diff
@@ -280,34 +342,34 @@ class RocksDBFileManager(
     val path = new Path(dfsRootDir)
 
     // All versions present in DFS, sorted
-    val sortedVersions = fm.list(path, onlyZipFiles)
+    val sortedSnapshotVersions = fm.list(path, onlyZipFiles)
       .map(_.getPath.getName.stripSuffix(".zip"))
       .map(_.toLong)
       .sorted
 
     // Return if no versions generated yet
-    if (sortedVersions.isEmpty) return
+    if (sortedSnapshotVersions.isEmpty) return
 
     // Find the versions to delete
-    val maxVersionPresent = sortedVersions.last
-    val minVersionPresent = sortedVersions.head
-    val minVersionToRetain =
-      math.max(minVersionPresent, maxVersionPresent - numVersionsToRetain + 1)
-    val versionsToDelete = sortedVersions.takeWhile(_ < minVersionToRetain).toSet[Long]
+    val maxSnapshotVersionPresent = sortedSnapshotVersions.last
+    val minSnapshotVersionPresent = sortedSnapshotVersions.head
+
+    // In order to reconstruct numVersionsToRetain version, retain the latest snapshot
+    // that satisfies (version <= maxSnapshotVersionPresent - numVersionsToRetain + 1)
+    val minVersionToRetain = sortedSnapshotVersions
```

Comment:
I'd say this is confusing. Technically saying, this logic by itself is arguably incorrect at least based on the name of the variable (I was going to explain the defect), but below snapshotVersionsToDelete logic covers it.

I'd say, use -1 as default value of foldLeft, and check minVersionToRetain to return early if it's -1. That way at least minVersionToRetain is going to be correct. I agree it's more verbose but should be better to avoid confusion.

Reply:
I changed to use 0 as default value, retain all versions if we don't find any version out of range.